server/src/main/java/io/crate/planner/operators/NestedLoopJoin.java

```diff
@@ -194,26 +194,29 @@ public ExecutionPlan build(DependencyCarrier executor,
 
         LogicalPlan leftLogicalPlan = lhs;
         LogicalPlan rightLogicalPlan = rhs;
-        isDistributed = isDistributed &&
-                        (!left.resultDescription().nodeIds().isEmpty() && !right.resultDescription().nodeIds().isEmpty());
-        boolean blockNlPossible = !isDistributed && isBlockNlPossible(left, right);
-
         JoinType joinType = this.joinType;
+        boolean blockNlPossible = !isDistributed && isBlockNlPossible(left, right);
         boolean expectedRowsAvailable = lhs.numExpectedRows() != -1 && rhs.numExpectedRows() != -1;
-        if (expectedRowsAvailable && !orderByWasPushedDown && joinType.supportsInversion() &&
-            (isDistributed && lhs.numExpectedRows() < rhs.numExpectedRows() && orderByFromLeft == null) ||
-            (blockNlPossible && lhs.numExpectedRows() > rhs.numExpectedRows())) {
-            // 1) The right side is always broadcast-ed, so for performance reasons we switch the tables so that
-            //    the right table is the smaller (numOfRows). If left relation has a pushed-down OrderBy that needs
-            //    to be preserved, then the switch is not possible.
-            // 2) For block nested loop, the left side should always be smaller. Benchmarks have shown that the
-            //    performance decreases if the left side is much larger and no limit is applied.
-            ExecutionPlan tmpExecutionPlan = left;
-            left = right;
-            right = tmpExecutionPlan;
-            leftLogicalPlan = rhs;
-            rightLogicalPlan = lhs;
-            joinType = joinType.invert();
+
+        if (expectedRowsAvailable) {
+            isDistributed = isDistributed &&
+                            (!left.resultDescription().nodeIds().isEmpty() &&
+                             !right.resultDescription().nodeIds().isEmpty());
```

Comment:
I don't really understand this change. Why is the `isDistributed` value now only updated if the expected rows are available? It's used outside of the if block as well

Reply:
if `expectedRowsAvailable` is false, `(blockNlPossible && lhs.numExpectedRows() > rhs.numExpectedRows()))` can be still true, even the `numExpectedRows` are not valid and the whole expression becomes true and the swap happens, but is not correct. This happens only in a single node environment when a block nested-loops join is possible. I moved expectedRowsAvailable in a separate if statement to make sure the whole block is only executed when the rows are available. I moved is distributed inside the first if statement, because it needs only be evaluated when the expected rows are available.